src/security/context.cpp

```diff
@@ -870,94 +1008,524 @@ ngx_int_t Context::do_request_body_filter(ngx_http_request_t &request,
+    ngx_log_error(NGX_LOG_NOTICE, c.log, 0,
+                  "send_buffered_header: send_chain failed");
+    return NGX_ERROR;
+  }
+
+  ngx_log_debug(NGX_LOG_DEBUG_HTTP, c.log, 0,
+                "send_buffered_header: remaining chain_(len,size)=(%uz,%uz)",
+                chain_length(rem_chain), chain_size(rem_chain));
+  for (ngx_chain_t *cl = header_filter_ctx_.out; cl && cl != rem_chain;) {
+    ngx_chain_t *ln = cl;
+    cl = cl->next;
+    ngx_free_chain(c.pool, ln);
+  }
+  header_filter_ctx_.replace_out(rem_chain);
+  if (rem_chain == nullptr) {
+    return NGX_OK;
+  }
+  return NGX_AGAIN;
+}
+
 void Context::FilterCtx::clear(ngx_pool_t &pool) noexcept {
   for (ngx_chain_t *cl = out; cl;) {
     ngx_chain_t *ln = cl;
     cl = cl->next;
     ngx_free_chain(&pool, ln);
   }
+
   out = nullptr;
-  out_last = &out;
+  out_latest = &out;
+  out_total = 0;
+  copied_total = 0;
+  // found_last retained
 }
 
-ngx_int_t Context::do_output_body_filter(ngx_http_request_t &request,
-                                         ngx_chain_t *chain, dd::Span &span) {
+void Context::FilterCtx::replace_out(ngx_chain_t *new_out) noexcept {
+  out = new_out;
+  copied_total = 0;
+  out_total = 0;
+  ngx_chain_t **lastp = &out;
+  for (; *lastp; lastp = &(*lastp)->next) {
+    out_total += ngx_buf_size((*lastp)->buf);
+  }
+  out_latest = lastp;
+}
+
+namespace {
+class Http1TemporarySendChain {
+ public:
+  static Http1TemporarySendChain instance;
+  void activate(Context &ctx, ngx_http_request_t &request) noexcept {
+    current_ctx_ = &ctx;
+    prev_send_chain_ = request.connection->send_chain;
+    request.connection->send_chain = send_chain_save;
+  }
+
+  void deactivate(ngx_http_request_t &request) noexcept {
+    auto *ctx = instance.current_ctx_;
+    assert(ctx != nullptr);
+
+    if (request.out) {
+      // there is uncommitted data in the output chain; send_chain() was not
+      // called; ngx_http_write_filter() chose not to do it
+      ngx_log_debug(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                    "temporary send_chain: uncommitted data in output chain "
+                    "(%uz bytes); adding it to the header output buffer",
+                    chain_size(request.out));
+      ctx->buffer_header_output(*request.pool, request.out);
+      for (auto *cl = request.out; cl;) {
+        auto *next = cl->next;
+        ngx_free_chain(request.pool, cl);
+        cl = next;
+      }
+      request.out = nullptr;
+      request.connection->buffered &= ~NGX_HTTP_WRITE_BUFFERED;
+    }
+
+    current_ctx_ = nullptr;
+    request.connection->send_chain = prev_send_chain_;
+  }
+
+ private:
+  static ngx_chain_t *send_chain_save(ngx_connection_t *c, ngx_chain_t *in,
+                                      off_t limit) {
+    (void)limit;
+
+    ngx_http_request_t *req = static_cast<decltype(req)>(c->data);
+    assert(req != nullptr);
+
+    auto *ctx = instance.current_ctx_;
+    assert(ctx != nullptr);
```

Comment:
How certain are we about these assertions...? An explicit check may be more prudent otherwise.

Reply:
What's not clear is that c->data is the request in http3.

Seeing as this function is very difficult to exercise and the tests don't reach it (the data for the headers is small enough that it isn't sent just yet), I'm changing this to not rely on c->data.